pkg/services/annotations/annotationsimpl/loki/historian_store.go

```diff
@@ -48,28 +70,231 @@ func NewLokiHistorianStore(cfg setting.UnifiedAlertingStateHistorySettings, ft f
+	if err != nil {
+		return make([]*annotations.ItemDTO, 0), ErrLokiStoreInternal.Errorf("failed to build loki query: %w", err)
+	}
+
+	now := time.Now().UTC()
+	if query.To == 0 {
+		query.To = now.UnixMilli()
+	}
+	if query.From == 0 {
+		query.From = now.Add(-defaultQueryRange).UnixMilli()
+	}
+
+	// query.From and query.To are always in milliseconds, convert them to nanoseconds for loki
+	from := query.From * 1e6
+	to := query.To * 1e6
+
+	res, err := r.client.RangeQuery(ctx, logQL, from, to, query.Limit)
+	if err != nil {
+		return make([]*annotations.ItemDTO, 0), ErrLokiStoreInternal.Errorf("failed to query loki: %w", err)
+	}
+
+	items := make([]*annotations.ItemDTO, 0)
+	for _, stream := range res.Data.Result {
+		items = append(items, r.annotationsFromStream(stream, *accessResources)...)
+	}
+
+	// order by time desc
+	sort.Slice(items, func(i, j int) bool {
+		return items[i].Time > items[j].Time
+	})
+
+	return items, err
+}
+
+func (r *LokiHistorianStore) annotationsFromStream(stream historian.Stream, ac accesscontrol.AccessResources) []*annotations.ItemDTO {
+	items := make([]*annotations.ItemDTO, 0, len(stream.Values))
+	for _, sample := range stream.Values {
+		entry := historian.LokiEntry{}
+		err := json.Unmarshal([]byte(sample.V), &entry)
+		if err != nil {
+			// bad data, skip
+			continue
+		}
+
+		if !hasAccess(entry, ac) {
+			// no access to this annotation, skip
+			continue
+		}
+
+		transition, err := buildTransition(entry)
+		if err != nil {
+			// bad data, skip
+			continue
+		}
+
+		if !historian.ShouldRecordAnnotation(*transition) {
+			// skip non-annotation transition
+			continue
+		}
+
+		annotationText, annotationData := historian.BuildAnnotationTextAndData(
+			historymodel.RuleMeta{
+				Title: entry.RuleTitle,
+			},
+			transition.State,
+		)
+
+		items = append(items, &annotations.ItemDTO{
+			AlertID:      entry.RuleID,
+			DashboardID:  ac.Dashboards[entry.DashboardUID],
+			DashboardUID: &entry.DashboardUID,
+			PanelID:      entry.PanelID,
+			NewState:     entry.Current,
+			PrevState:    entry.Previous,
+			Time:         sample.T.UnixMilli(),
+			Text:         annotationText,
+			Data:         annotationData,
+		})
+	}
+
+	return items
 }
 
 func (r *LokiHistorianStore) GetTags(ctx context.Context, query *annotations.TagsQuery) (annotations.FindTagsResult, error) {
 	return annotations.FindTagsResult{}, nil
 }
 
+// util
+
+func getRule(ctx context.Context, sql db.DB, orgID int64, ruleID int64) (*ngmodels.AlertRule, error) {
+	rule := &ngmodels.AlertRule{OrgID: orgID, ID: ruleID}
+	err := sql.WithDbSession(ctx, func(sess *db.Session) error {
+		exists, err := sess.Get(rule)
+		if err != nil {
+			return err
+		}
+		if !exists {
+			return ErrMissingRule
+		}
+		return nil
+	})
+
+	return rule, err
+}
+
+func hasAccess(entry historian.LokiEntry, resources accesscontrol.AccessResources) bool {
+	if resources.CanAccessOrgAnnotations && entry.DashboardUID != "" {
```

Comment:
I believe this is incorrect. Is it supposed to be this?
```suggestion
	if !resources.CanAccessOrgAnnotations || entry.DashboardUID == "" {
```

Reply:
It is wrong yeah, to keep the short-circuit logic the correct statement would be this:

```suggestion
	if entry.DashboardUID == "" && !resources.CanAccessOrgAnnotations {
```